opentelemetry/src/context.rs

```diff
@@ -381,6 +389,75 @@ impl Hasher for IdHasher {
     }
 }
 
+struct ContextStack {
+    current_cx: Context,
+    current_id: usize,
+    // TODO:ban wrap the whole id thing in its own type
+    id_count: usize,
+    // TODO:ban wrap the the tuple in its own type
+    stack: Vec<Option<(usize, Context)>>,
+}
+
+impl ContextStack {
+    #[inline(always)]
+    fn push(&mut self, cx: Context) -> usize {
+        self.id_count += 512; // TODO:ban clean up this
+        let next_id = self.stack.len() + 1 + self.id_count;
+        let current_cx = std::mem::replace(&mut self.current_cx, cx);
+        self.stack.push(Some((self.current_id, current_cx)));
+        self.current_id = next_id;
+        next_id
+    }
+
+    #[inline(always)]
+    fn pop_id(&mut self, id: usize) {
+        if id == 0 {
+            return;
+        }
+        // Are we at the top of the stack?
+        if id == self.current_id {
+            // Shrink the stack if possible
+            while let Some(None) = self.stack.last() {
+                self.stack.pop();
+            }
+            // There is always the initial context at the bottom of the stack
+            if let Some(Some((next_id, next_cx))) = self.stack.pop() {
+                self.current_cx = next_cx;
+                self.current_id = next_id;
+            }
+        } else {
+            let pos = id & 511; // TODO:ban clean up this
```

Comment:
Can there be collision here - different id's mapping to same pos, and the logic mistakenly trying to remove the wrong entry?

Reply:
Seems the increment by 512 in push is trying to avoid the collision, but I am not sure if it prevents entirely.

Reply:
Thinking some more about it, I actually think that there is no need for anything other than the stack position in the `ContextGuard`, since it's `!Send` and doesn't implement `Copy` or `Clone`. The `ContextStack` is thread local, and the `ContextGuard` is limited to the same thread. A stack position will not be reused until it is empty, so there shouldn't be any collisions.